pkg/controllers/trace_controller.go

```diff
@@ -193,33 +239,30 @@ func (r *TraceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl
 		return ctrl.Result{}, err
 	}
 
-	// Call gadget operation
-	traceBeforeOperation := trace.DeepCopy()
-	patch := client.MergeFrom(traceBeforeOperation)
-
+	// Check operation is supported for this specific gadget
 	gadgetOperation, ok := factory.Operations()[op]
 	if !ok {
-		trace.Status.OperationError = fmt.Sprintf("Unknown operation %q", op)
-	} else {
-		gadgetOperation.Operation(req.NamespacedName.String(), trace)
+		setTraceOpError(ctx, r.Client, req.NamespacedName.String(),
+			trace, fmt.Sprintf("Unsupported operation %q for gadget %q",
+				op, trace.Spec.Gadget))
+
+		return ctrl.Result{}, nil
 	}
 
+	// Call gadget operation
+	traceBeforeOperation := trace.DeepCopy()
+	patch := client.MergeFrom(traceBeforeOperation)
+	gadgetOperation.Operation(req.NamespacedName.String(), trace)
+
 	if apiequality.Semantic.DeepEqual(traceBeforeOperation.Status, trace.Status) {
 		log.Info("Gadget completed operation without changing the trace status")
-	} else {
-		log.Infof("Gadget completed operation: updating state=%s operationError=%s output=<%d characters>",
-			trace.Status.State,
-			trace.Status.OperationError,
-			len(trace.Status.Output),
-		)
-		err = r.Client.Status().Patch(ctx, trace, patch)
-		if err != nil {
-			log.Errorf("Failed to update trace status: %s", err)
-			return ctrl.Result{}, err
-		}
+		return ctrl.Result{}, nil
 	}
 
-	return ctrl.Result{}, nil
+	log.Infof("Gadget completed operation. Trace status will be updated accordingly")
+
+	return ctrl.Result{}, updateTraceStatus(ctx, r.Client,
```

Comment:
Here you still potentially return an error. Is it on purpose?

Reply:
Yeah, I kept it like this because I didn't want to change the current behaviour. However, now that you highlighted it, I think it is not useful to ask the controller to call the Reconcile again because the annotation was already removed so it will do nothing.

If you agree, I will update the PR to not return an error here.

Reply:
Agree, let's return `nil`.